light/beacon/binary_merkle.go

```diff
@@ -0,0 +1,555 @@
+	if !ok {
+		return common.Hash{}, false
+	}
+	rhash, ok := TraverseProof(rr, wr)
+	if !ok {
+		return common.Hash{}, false
+	}
+	if !nodeAvailable {
+		hasher := sha256.New()
+		hasher.Write(lhash[:])
+		hasher.Write(rhash[:])
+		hasher.Sum(node[:0])
+	}
+	if writer != nil {
+		writer.writeNode(node)
+	}
+	return common.Hash(node), true
+}
+
+// MultiProof stores a partial Merkle tree proof
+type MultiProof struct {
+	Format ProofFormat
+	Values MerkleValues
+}
+
+// ParseSSZMultiProof creates a MultiProof from a standard serialized format
+// (2 byte subtree sizes, then 32 byte leaf values and internal sibling hashes
+// in depth-first, left-to-right order).
+func ParseSSZMultiProof(proof []byte) (MultiProof, error) {
+	if len(proof) < 3 || proof[0] != 1 {
+		return MultiProof{}, errors.New("invalid proof length")
+	}
+	var (
+		leafCount   = int(binary.LittleEndian.Uint16(proof[1:3]))
+		format      = NewIndexMapFormat()
+		values      = make(MerkleValues, leafCount)
+		valuesStart = leafCount*2 + 1
+	)
+	if len(proof) != leafCount*34+1 {
+		return MultiProof{}, errors.New("invalid proof length")
+	}
+	if err := parseMultiProofFormat(format.leaves, 1, proof[3:valuesStart]); err != nil {
+		return MultiProof{}, err
+	}
+	for i := range values {
+		copy(values[i][:], proof[valuesStart+i*32:valuesStart+(i+1)*32])
+	}
+	return MultiProof{Format: format, Values: values}, nil
+}
+
+// parseMultiProofFormat recursively parses the SSZ serialized proof format
+func parseMultiProofFormat(leaves map[uint64]ProofFormat, index uint64, format []byte) error {
+	if len(format) == 0 {
+		return nil
+	}
+	leaves[index] = nil
+	boundary := int(binary.LittleEndian.Uint16(format[:2])) * 2
+	if boundary > len(format) {
+		return errors.New("invalid proof format")
+	}
+	if err := parseMultiProofFormat(leaves, index*2, format[2:boundary]); err != nil {
+		return err
+	}
+	if err := parseMultiProofFormat(leaves, index*2+1, format[boundary:]); err != nil {
+		return err
+	}
+	return nil
+}
+
+// multiProofReader implements ProofReader based on a MultiProof and also allows
+// attaching further subtree readers at certain indices
+type multiProofReader struct {
+	format   ProofFormat
+	values   *MerkleValues
+	index    uint64
+	subtrees func(uint64) ProofReader
+}
+
+// children implements ProofReader
+func (mpr multiProofReader) children() (left, right ProofReader) {
+	lf, rf := mpr.format.children()
+	if lf == nil {
+		if mpr.subtrees != nil {
+			if subtree := mpr.subtrees(mpr.index); subtree != nil {
+				return subtree.children()
+			}
+		}
+		return nil, nil
+	}
+	return multiProofReader{format: lf, values: mpr.values, index: mpr.index * 2, subtrees: mpr.subtrees},
+		multiProofReader{format: rf, values: mpr.values, index: mpr.index*2 + 1, subtrees: mpr.subtrees}
+}
+
+// readNode implements ProofReader
+func (mpr multiProofReader) readNode() (MerkleValue, bool) {
+	if l, _ := mpr.format.children(); l == nil && len(*mpr.values) > 0 {
+		hash := (*mpr.values)[0]
```

Comment:
Maybe we can maintain a pointer inside of multiProofReader.
pointer = len(values) if it's exhausted.

Reply:
I did the change, though the pointer still needs to be a `*int` reference for the same reason as in the case of `multiProofWriter`. Even so, it probably looks nicer than devouring the slice from its head :)